Feature/cdap 12017 metrics salting

CuriousVini opened the pull request:
Issue: https://issues.cask.co/browse/CDAP-12017
Build: https://builds.cask.co/browse/CDAP-DUT5937-18
ITN: `StreamTest` and `WordCountTest` which has metrics validations are passing on cluster.

Tested Metrics on cluster.

Remaining:
- Performance tests 
- Upgrade tests

CuriousVini committed:
Adding salting to metrics table

anwar6953 commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java:
```diff
@@ -112,9 +119,88 @@ private MetricsTable getOrCreateMetricsTable(String tableName, DatasetProperties
+    return getSplitKeys(splits, splits, rowKeyDistributor);
+  }
+
+  public static byte[][] getSplitKeys(int splits, int buckets, AbstractRowKeyDistributor keyDistributor) {
```
why must this be public? Seems like its only used within this class.
And it seems like `splits always equals buckets`, from the single usage of this method?

anwar6953 commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseScanner.java:
```diff
@@ -32,10 +34,19 @@
+  }
 
   public HBaseScanner(ResultScanner scanner, byte[] columnFamily) {
     this.scanner = scanner;
```
nit - can be made 1liner: `HBaseScanner(scanner, columnFamily, null);`

anwar6953 commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseMetricsTable.java:
```diff
@@ -71,11 +94,49 @@ public HBaseMetricsTable(DatasetContext datasetContext, DatasetSpecification spe
+        public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+          REJECTION_LOG.info(
+            "No more threads in the HBase scan thread pool. Consider increase {}. Scan from caller thread {}",
+            Constants.MessagingSystem.HBASE_MAX_SCAN_THREADS, Thread.currentThread().getName()
```
seems like you're overloading a property that has 'messaging' in its name. This can be misleading.
Probably have your own property name, even if the behavior is similar. Or if the common use case is that the value of both would need updating often, rename the property to something that isn't specific to messaging, and then reuse it.
Edit: nevermind, you are using your own property name. But this LOG.info message still references the wrong property.

anwar6953 replied:
Also, seems like you're getting it from the dataset spec. Shouldn't it get it from the cConf instead?

anwar6953 commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -0,0 +1,110 @@
+      for (byte[] column : entry.getValue().keySet()) {
+        columns.add(column);
+      }
+      // only delete from old table
```
why not delete from new table?

CuriousVini replied:
This is because when we put we delete the column from old table only for gauge metrics. So that when we read both tables we can directly sum them up. This is the way to differentiate gauge and count metrics

anew replied:
You should explain that in a comment in the code. The current comment is not very useful.

anwar6953 commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java:
```diff
@@ -157,7 +157,7 @@ private MetricsTable getOrCreateResolutionMetricsTable(String v2TableName, Strin
   }
 
-  public static byte[][] getSplitKeys(int splits, int buckets, AbstractRowKeyDistributor keyDistributor) {
+  private static byte[][] getSplitKeys(int splits, int buckets, AbstractRowKeyDistributor keyDistributor) {
```
why is splits always equal to buckets?
Do we need to pass them in as separate variables?
Or is it the case that we will always have splits==buckets for our current use case, and we may expand beyond this in the future?

CuriousVini replied:
so that the key distribution is even across number of splits

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java:
```diff
@@ -585,7 +585,16 @@
+    public static final String DEFAULT_METRIC_V3_TABLE_PREFIX = "metrics.v3.table";
+    public static final String METRICS_HBASE_MAX_SCAN_THREADS = "metrics.hbase.max.scan.threads";
+    public static final String METICS_HBASE_TABLE_SPLITS = "metrics.v3.table.splits";
+    // Hardcode this value because we do not want user to control it
```
so it is not really hardcode, as it still get it through cconf, right?

anwar6953 replied:
I dont think it should be in the dataset properties though. Only in the cconf, imo.
Edit: sorry, I was referring to the `METRICS_HBASE_MAX_SCAN_THREADS`. That shouldn't be in the dataset properties.

anew replied:
typo: METRICS

CuriousVini commented on cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java:
```diff
@@ -585,7 +585,16 @@
+    public static final String DEFAULT_METRIC_V3_TABLE_PREFIX = "metrics.v3.table";
+    public static final String METRICS_HBASE_MAX_SCAN_THREADS = "metrics.hbase.max.scan.threads";
+    public static final String METICS_HBASE_TABLE_SPLITS = "metrics.v3.table.splits";
+    // Hardcode this value because we do not want user to control it
```
Why not use as dataset properties @anwar6953 ?

anew commented on cdap-data-fabric-tests/src/test/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseMetricsTableTest.java:
```diff
@@ -147,11 +153,97 @@ public void testConcurrentIncrement() throws Exception {
     DatasetId metricsDatasetInstanceId = NamespaceId.SYSTEM.dataset(name);
-    DatasetProperties props = TableProperties.builder().setReadlessIncrementSupport(true).build();
+    DatasetProperties props = TableProperties.builder().setReadlessIncrementSupport(true)
+      .add(Constants.Metrics.METICS_HBASE_TABLE_SPLITS, 16).build();
```
Should this use the constant METRICS_HBASE_SPLITS instead of 16?

anew replied:
We actually want to pre-split the table, too, right? Where is that being done?

anew replied:
I realize now that this is in a test case. The test case should create the table in the same way that the metrics system would. Ideally call that method from the test case.

CuriousVini replied:
Removed splits property because it is not needed for tests.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseMetricsTable.java:
```diff
@@ -71,11 +94,49 @@ public HBaseMetricsTable(DatasetContext datasetContext, DatasetSpecification spe
+                                                 Threads.createDaemonThreadFactory("metrics-hbase-scanner-%d"),
+                                                 callerRunsPolicy);
+
+      int splits = spec.getIntProperty(Constants.Metrics.METICS_HBASE_TABLE_SPLITS,
```
So storing this in the dataset properties and retrieving it here means that it is not hard coded? What happens if it gets updated , say from 16 to 32? Does it make sense to have it as a property?

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -0,0 +1,110 @@
+    for (Map.Entry<byte[], ? extends SortedMap<byte[], Long>> entry : updates.entrySet()) {
+      byte[] row = entry.getKey();
+      List<byte[]> columns = Lists.newArrayList();
+      for (byte[] column : entry.getValue().keySet()) {
```
Why copy to a list first and then call toArray()? You can call toArray() directly on the key set.
```
Set<byte[]> columnSet = entry.getValue().keySet();
byte[][] columns = columnSet.toArray(new byte[columnSet.size()][]);
v2HBaseTable.delete(row, columns);
```

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  @Override
+  public void put(SortedMap<byte[], ? extends SortedMap<byte[], Long>> updates) {
```
is used only for gauges, or also for counters?

CuriousVini replied:
This is only used for gauges.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -0,0 +1,110 @@
+import javax.annotation.Nullable;
+
+/**
+ * Combined Metrics table to read from both v2 and v3 Tables
```
This Javadoc should explain the logic for merging:
- when reading
- when writing; why in some cases we have to delete from the old table

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  @Override
+  public boolean swap(byte[] row, byte[] column, byte[] oldValue, byte[] newValue) {
```
what is this used for? Can it happen that the existing value is in the old table?

CuriousVini replied:
It is used in EntityTable to recycle entity ids: https://github.com/caskdata/cdap/blob/develop/cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/timeseries/EntityTable.java#L154

No usage related to resolution tables.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedMetricsScanner.java:
```diff
@@ -0,0 +1,178 @@
+
+    // v2 scanner is exhausted, so return row from v3 table
+    if (v2TableNextRow == null) {
+      return getV3Row();
```
the name of the method (getV3Row) suggests that it only gets. But it also advances the v3 table scanner and replaces the nextV3Row with the next result. It would be better to name the method something like advanceV3Row or advanceV3Scanner.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedMetricsScanner.java:
```diff
@@ -0,0 +1,178 @@
+    int rowComparison = Bytes.compareTo(v2TableNextRow.getRow(), v3TableNextRow.getRow());
+
+    if (rowComparison > 0) {
+      resultRow = getV3Row();
```
return getV3Row();

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedMetricsScanner.java:
```diff
@@ -0,0 +1,178 @@
+    if (rowComparison > 0) {
+      resultRow = getV3Row();
+    } else if (rowComparison < 0) {
+      resultRow = getV2Row();
```
return getV2Row(); 

then the rest does not need to be inside else { }

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedMetricsScanner.java:
```diff
@@ -0,0 +1,178 @@
+
+      mergeColumns(v2TableNextRow, v3TableNextRow, mergedColumns);
+
+      v2TableNextRow = v2TableScanner.next();
```
you have a helper method getV2Row() - that I suggested to rename. It would be better to use that here.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedMetricsScanner.java:
```diff
@@ -0,0 +1,178 @@
+
+      v2TableNextRow = v2TableScanner.next();
+      v3TableNextRow = v3TableScanner.next();
+      resultRow = new Result(rowKey, mergedColumns);
```
return new Result(...);

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedMetricsScanner.java:
```diff
@@ -0,0 +1,178 @@
+  private void mergeColumns(Row v2TableNextRow, Row v3TableNextRow, Map<byte[], byte[]> mergedColumns) {
+    Iterator<Map.Entry<byte[], byte[]>> v2ColumnIterator = v2TableNextRow.getColumns().entrySet().iterator();
+    Iterator<Map.Entry<byte[], byte[]>> v3ColumnIterator = v3TableNextRow.getColumns().entrySet().iterator();
+
```
seems that it would be a lot simpler to 
1. copy v2 row into a new Map
2. iterate over v3 row and for each column:
- if it already exists, convert both values to long, add them and write the sum back as bytes
- otherwise copy this column to the new map.

CuriousVini replied:
Thanks! It is much simpler than what I had.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseMetricsTable.java:
```diff
@@ -71,11 +94,49 @@ public HBaseMetricsTable(DatasetContext datasetContext, DatasetSpecification spe
+        @Override
+        public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+          REJECTION_LOG.info(
+            "No more threads in the HBase scan thread pool. Consider increase {}. Scan from caller thread {}",
```
increasing

anew replied:
What does "Scan from caller thread" mean to the user? I think it is less confusing to say "Caller thread ID is {}".

anew replied:
Oh wait, you are trying to say: because the thread pool is exhausted, the scan is performed synchronously in the  caller thread. Then perhaps phrase it as "Performing scan in caller thread {}".

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseMetricsTable.java:
```diff
@@ -156,12 +221,16 @@ public void increment(byte[] row, Map<byte[], Long> increments) {
     }
   }
 
+  private byte[] getRowKey(byte[] row) {
```
rename to createDistributedRowKey or createActualRowKey

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseScanner.java:
```diff
@@ -55,13 +63,13 @@ public Row next() {
         Map<byte[], byte[]> rowMap = HBaseTable.getRowMap(result, columnFamily);
         if (rowMap.size() > 0) {
-          return new co.cask.cdap.api.dataset.table.Result(result.getRow(), rowMap);
+          byte[] row = rowKeyDistributor == null ? result.getRow() : rowKeyDistributor.getOriginalKey(result.getRow());
```
is the rowKey always needed by the caller? If not, you could consider using a Result object that calls getOriginalKey() only if getRow() is called.

CuriousVini replied:
The Rowkey is used by `FactScanner` to get measure name from the row key to create `FactScanResult`

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java:
```diff
@@ -112,9 +119,88 @@ private MetricsTable getOrCreateMetricsTable(String tableName, DatasetProperties
+
+      MetricsTable v2Table = null;
+      if (dsFramework.hasInstance(v2TableId)) {
+        v2Table = DatasetsUtil.getOrCreateDataset(dsFramework, v2TableId, MetricsTable.class.getName(),
```
no need to getOrCreate() if it already exists?

CuriousVini replied:
It v2 table dataset exists, we get v2Table instance so that we can pass it to `CombinedMetricsTable`. Any better way to get `v2Table` dataset instance?

anew replied:
Yes, but you are creating a race condition here. If the migration finishes, it will delete the v2 Table. That can happen in between the hasInstance() and the getOrCreate(). This would then recreate the v2Table. It is better to simply use datasetFramework.getDataset(). If the table exists, it will return non-null.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java:
```diff
@@ -112,9 +119,88 @@ private MetricsTable getOrCreateMetricsTable(String tableName, DatasetProperties
+    return getSplitKeys(splits, splits, rowKeyDistributor);
+  }
+
+  private static byte[][] getSplitKeys(int splits, int buckets, AbstractRowKeyDistributor keyDistributor) {
```
seems that this should be a method of the rowkeyDistributor. As it relies on knowledge about how salt keys are formed.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseMetricsTable.java:
```diff
@@ -260,13 +337,26 @@ private ScanBuilder configureRangeScan(ScanBuilder scan, @Nullable byte[] startR
   }
 
+  private void saltFilterPairs(List<Pair<byte[], byte[]>> fuzzyPairs, ImmutablePair<byte[], byte[]> pair) {
+    byte[][] firstAllDistKeys = rowKeyDistributor.getAllDistributedKeys(pair.getFirst());
```
this should be a method of the rowkey distributor. As it relies on its internals.

anew reviewed (changes requested):
Please see my comments.

CuriousVini commented:
Thanks @anew for the review. I have addressed comment. Please take another pass when you get a chance!
Tested latest changes on cluster with v3 tables.

CuriousVini commented:
Build is green https://builds.cask.co/browse/CDAP-DUT5937-25

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -19,17 +19,27 @@
+ * Reading:
+ * When we scan, we scan from both v2 and v3 tables using {@link CombinedMetricsScanner}. For higher resolution
+ * tables, if the same row is divided between v2 and v3 tables, we merge the columns and add the values of
+ * overlapping columns.
```
Add: this also works for gauges, because when writing to the v3 table, we delete the corresponding entry in v2 table. Therefore we will never encounter a value in both tables for a gauge.

anew commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/hbase/wd/AbstractRowKeyDistributor.java:
```diff
@@ -87,4 +91,57 @@ public void addInfo(Configuration conf) {
+    return fuzzyPairs;
+  }
+
+  public byte[][] getSplitKeys(int splits, int buckets) {
```
missing javadoc

anew commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/hbase/wd/AbstractRowKeyDistributor.java:
```diff
@@ -87,4 +91,57 @@ public void addInfo(Configuration conf) {
     }
   }
+
+  public List<Pair<byte[], byte[]>> getDistributedFilterPairs(ImmutablePair<byte[], byte[]> pair) {
```
missing javadoc

anew commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/hbase/wd/AbstractRowKeyDistributor.java:
```diff
@@ -15,14 +15,18 @@
 package co.cask.cdap.hbase.wd;
 
+import co.cask.cdap.common.utils.ImmutablePair;
+import com.google.common.base.Preconditions;
```
please update the copyright to 2017

anew commented on cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java:
```diff
@@ -584,7 +584,6 @@
     // NOTE: "v3" to avoid conflict with data of older metrics system
     public static final String DEFAULT_METRIC_V3_TABLE_PREFIX = "metrics.v3.table";
     public static final String METRICS_HBASE_MAX_SCAN_THREADS = "metrics.hbase.max.scan.threads";
-    public static final String METICS_HBASE_TABLE_SPLITS = "metrics.v3.table.splits";
```
On second thought, this number should be configurable in the cConf. Larger clusters may need more than 16. I am not sure how this can be changed, though, once the tables are created. Sp perhaps we can get it from cConf when creating the table, store it in the table properties, and at runtime use the number from the properties. Sorry that is just how you had it initially. But we have to make sure the dataset properties cannot be changed, so MetricsTableDefinition.reconfigure() must throw an exception if it changes. 

This does not have to be in this PR, can you open a Jira for making metrics table splits configurable?

CuriousVini replied:
Sure. https://issues.cask.co/browse/CDAP-12297

anew reviewed (changes requested):
Thanks for addressing my comments. I only have a handful of additional comments.

CuriousVini commented:
Thanks @anew I have addressed comments. Please take another pass when you get a chance.

anew commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/hbase/wd/AbstractRowKeyDistributor.java:
```diff
@@ -92,6 +92,13 @@ public void addInfo(Configuration conf) {
 
+  /**
+   * Method to salt the row key of filter pair. It also adds a new byte with value 1 which means that this byte
+   * in provided row key is NOT fixed
```
add missing period at the end. Does this pass checkstyle?

anew commented on cdap-hbase-compat-base/src/main/java/co/cask/cdap/hbase/wd/AbstractRowKeyDistributor.java:
```diff
@@ -103,6 +110,13 @@ public void addInfo(Configuration conf) {
   }
 
+  /**
+   * Get all the split keys based on splits and number of buckets
```
Same here, add a period.

anew reviewed (commented):
Only two minor comments on style. Please fix, squash and and merge without further review.

CuriousVini commented:
Resolved conflict in `cdap-default.xml` because of https://github.com/caskdata/cdap/pull/9358/files#diff-d2f273e30d51501da90e3f8cea5929e0R2053. Merging now.
